### DISCUSSION TITLE
How to call a rake task with good_job.cron?

### DISCUSSION THREAD

**Discussion by steveroot** [bensheldon/good_job#1039] [Q&A] [answered]
Hi, I'm playing with an app that I want to send an email reminder X time before an event.
I have a rake task "send_meeting_email" that works out when to send them.
I know goodjob is working as I can see in the dashboard emails queued from a controller using deliver_later.

How do I call a rake task please?

I have this in production.rb. 2 min interval for whilst I test and I guessed class was simply going to be the rake task name, but I guessed wrong :-)

Thanks
Steve

```
  # Configure cron with a hash that has a unique key for each recurring job
  config.good_job.cron = {
    # Every 15 minutes, enqueue `ExampleJob.set(priority: -10).perform_later(52, name: "Alice")`
    frequent_task: { # each recurring job must have a unique key
      cron: "*/2 * * * *", # 2 minutes to test cron-style scheduling format by fugit gem
      class: "send_meeting_email", # reference the Job class with a string
      #args: [42, { name: "Alice" }], # arguments to pass; can also be a proc
      #set: { priority: -10 }, # additional ActiveJob properties; can also be a lambda/proc
      description: "rake tast to see if emails need to be sent", # optional description
    },
    #another_task: {
    #  cron: "0 0,12 * * *",
    #  class: "AnotherJob",
    #},
    # etc.
  }
```

**Comment by bensheldon** [comment, marked as answer]
Sorry about the confusion! The cron `class` must be the name of an Active Job job. You'll need to create a job that runs the Rake task. 

Rake can be a little tricky to run in arbitrary ruby code, so if it's possible, my recommendation would be to extract the behavior from the Rake task into its own Ruby class (e.g. `SendMeetingEmail.new.send`) and then call that in _both_ your Rake task and a Job. Or you could even extract all of it to a job and then perform_now it in the Rake task (`SendMeetingEmailJob.perform_now`).

If you're using some gem/library in which it's not possible to extract or rewrite the Rake task. In which case, you can make a job that invokes the rake task. There's a lot variables as to whether to `invoke/reenable` or `execute` the task but the simplest implementation would be:


```ruby
class SendEmailJob < ApplicationJob
  def perform
    Rails.application.load_tasks
    Rake::Task['send_meeting_email'].invoke
    Rake::Task['send_meeting_email'].reenable # necessary so the rake task can run multiple times in the same process
  end
end
```

**Reply by steveroot** [reply]
Thank you Ben, 
No need to apologise, that's amazing! I don't even need the rake task now
